include/repeated_hafnian.hpp

```diff
@@ -16,215 +16,303 @@
  * \rst
  * Contains functions for computing the hafnian using the algorithm
  * described in *From moments of sum to moments of product*,
- * `doi:10.1016/j.jmva.2007.01.013 <https://dx.doi.org/10.1016/j.jmva.2007.01.013>`__.
- * \endrst
+ * `doi:10.1016/j.jmva.2007.01.013
+ * <https://dx.doi.org/10.1016/j.jmva.2007.01.013>`__. \endrst
  */
 #pragma once
 #include <stdafx.h>
 
 namespace libwalrus {
 
+/**
+ * Converts a linear index to a multi index
+ * e.g. if we wanted the multi-index (i,j) of an element in a 2x2 matrix given
+ * a linear index of 3 in the array storing the matrix, the maxes vector would
+ * be {1,1} and this function would return (1,0)
+ *
+ * @param linear_index the "flattened" index
+ * @param maxes a vector of integers, representing the max index value
+ * of each indice in the multi-index object.
+ *
+ * @return hafnian of the input matrix
+ */
+
+std::vector<int> lin_to_multi(unsigned long long int linear_index, const std::vector<int> &maxes) {
+  std::vector<int> i(maxes.size(), 0);
+  unsigned int l = 0;
+  int s0 = maxes[0] + 1;
+  i[l] = linear_index % s0;
+  unsigned long long int prod = s0;
+  unsigned long long int gl = linear_index;
+
+  while (l < i.size() - 1) {
+    unsigned long long int glp1 = (gl - i[l]) / (maxes[l] + 1);
+    i[l + 1] = glp1 % (maxes[l + 1] + 1);
+    prod *= (maxes[1 + 1] + 1);
+    gl = glp1;
+    l++;
+  }
+  return i;
+}
+
+/**
+ * Returns the binomial coefficient N!/K!(N-K)!
+ * Adapted from http://blog.plover.com/math/choose.html
+ *
+ * @param N
+ * @param K
+ *
+ * @return N!/K!(N-K)!
+ */
+
+template <typename T> T get_binom_coeff(T N, T K) {
+  T r = 1;
+  T d;
+  if (K > N)
+    return 0;
+  for (d = 1; d <= K; d++) {
+    r *= N--;
+    r /= d;
+  }
+  return r;
+}
+
 /**
  * Returns the hafnian of a matrix using the algorithm
  * described in *From moments of sum to moments of product*,
  * [doi:10.1016/j.jmva.2007.01.013](https://dx.doi.org/10.1016/j.jmva.2007.01.013>).
  *
- * Note that this algorithm, while generally slower than others, can be significantly more
- * efficient in the cases where the matrix has repeated rows and columns.
+ * Note that this algorithm, while generally slower than others, can be
+ * significantly more efficient in the cases where the matrix has repeated rows
+ * and columns.
  *
  * @param mat a flattened vector of size \f$n^2\f$, representing an
  *      \f$n\times n\f$ row-ordered symmetric matrix.
  * @param rpt a vector of integers, representing the number of
  *      times each row/column in `mat` is repeated. For example,
- *      `mat = [1]` and `rpt = [6]` represents a \f$6\times 6\f$ matrix of all ones.
+ *      `mat = [1]` and `rpt = [6]` represents a \f$6\times 6\f$ matrix of all
+ * ones.
```

Comment:
Does this need to be indented?

Reply:
no I'll fix that, it was indented for some reason by clang-format